hazelcast/src/main/java/com/hazelcast/internal/probing/ProbeRegistry.java

```diff
@@ -0,0 +1,107 @@
+package com.hazelcast.internal.probing;
+
+import com.hazelcast.internal.metrics.Probe;
+import com.hazelcast.internal.metrics.ProbeLevel;
+import com.hazelcast.spi.annotation.Beta;
+import com.hazelcast.spi.annotation.PrivateApi;
+import com.hazelcast.spi.impl.servicemanager.ServiceManager;
+
+/**
+ * A service made accessible to core services so they have a chance to register
+ * the "root objects" of the system.
+ * 
+ * After the registration phase the set of this {@link ProbeSource}s is fix.
+ * There is no need to unregister them later on as they always have the choice
+ * to stop rendering something.
+ * 
+ * Metrics about things that change over time require a corresponding
+ * {@link ProbeSource} (existing since the startup) that does probe the
+ * available instances as they appear and stops doing that as they disappear.
+ */
+@Beta
+@PrivateApi
+public interface ProbeRegistry {
+
+    /**
+     * Called once at startup, typically by a core service registering itself.
+     * 
+     * @param source a object that "knows" how to render metrics in its context
+     */
+    void register(ProbeSource source);
+
+    /**
+     * Legacy support instances only known by their interface with implementations
+     * that possibly implement {@link ProbeSource}.
+     * 
+     * @param source a object possibly implementing {@link ProbeSource}, might be
+     *        null as well
+     */
+    void registerIfSource(Object source);
+
+    /**
+     * @return Creates a new "private "context that should be kept by the caller to
+     *         render the contents of this {@link ProbeRegistry}. The implementation
+     *         will not support multi-threading as each thread should create its own
+     *         context instance.
+     */
+    ProbeRenderContext newRenderingContext();
```

Comment:
Can't the method below be implemented as method on the ProbeRenderContext?

ProbeRenderContext ctx = registry.newRenderingContext().filter(...);

Also there is a mismatch in the names. ProbeRENDERContext and newRENDERINGContext()

Reply:
Both ways are possible but this way has larger freedom on how to implement it while the other limits it quite a lot. Also I would like to keep `ProbeRenderContext` a _functional interface_. We might get Java 8 some day.

Reply:
Fixed the naming.